source/unified-test-format/schema-1.12.json

```diff
@@ -0,0 +1,692 @@
+            "additionalProperties": false,
+            "properties": {}
+          },
+          "connectionClosedEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {
+              "reason": { "type": "string" }
+            }
+          },
+          "connectionCheckOutStartedEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {}
+          },
+          "connectionCheckOutFailedEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {
+              "reason": { "type": "string" }
+            }
+          },
+          "connectionCheckedOutEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {}
+          },
+          "connectionCheckedInEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {}
+          }
+        }
+      },
+  
+      "expectedSdamEvent": {
+        "type": "object",
+        "additionalProperties": false,
+        "maxProperties": 1,
+        "minProperties": 1,
+        "properties": {
+          "serverDescriptionChangedEvent": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {
+              "previousDescription": { "$ref": "#/definitions/serverDescription" },
+              "newDescription": { "$ref": "#/definitions/serverDescription" }
+            }
+          }
+        }
+      },
+  
+      "serverDescription": {
+        "type": "object",
+        "additionalProperties": false,
+        "properties": {
+          "type": { 
+            "type": "string",
+            "enum": [
+              "Standalone",
+              "Mongos",
+              "PossiblePrimary",
+              "RSPrimary",
+              "RSSecondary",
+              "RSOther",
+              "RSArbiter",
+              "RSGhost",
+              "LoadBalancer",
+              "Unknown"
+            ]
+          }
+        }
+      },
+
+      "expectedLogMessagesForClient": {
+        "type": "object",
+        "additionalProperties": false,
+        "required": ["client", "messages"],
+        "properties": {
+          "client": { "type": "string" },
+          "messages": {
+            "type": "array",
+            "items": { "$ref": "#/definitions/expectedLogMessage" } 
+            }
+        }
+      },
+
+      "expectedLogMessage": {
+        "type": "object",
+        "additionalProperties": false,
+        "required": ["level", "component", "data"],
+        "properties": {
+            "level": { "$ref": "#/definitions/logSeverityLevel" },
+            "component": { "$ref": "#/definitions/logComponent" },
+            "data": { "type": "object" },
+            "hasFailure": {"type": "boolean" },
+            "failureIsRedacted": { "type": "boolean" }
+        },
+        "anyOf": [
```

Comment:
paging @jmikola. what I'm trying to express here is that any of the following are valid:
* specifying neither `hasFailure` nor `failureIsRedacted`
* specifying `hasFailure` (either true or false) and not specifying `failureIsRedacted`
* specifying `hasFailure=true` and `failureIsRedacted` (either true or false)

but that it is disallowed to specify `failureIsRedacted` without `hasFailure=true`. 

relatedly, I debated whether it made sense to allow `hasFailure` without `failureIsRedacted`, since it is redundant with specifying `failure: { $$exists: true }` in `data`. I went with allowing it for now, but if you think it would be better to only allow it along with `failureIsRedacted` I have no objections to being more restrictive. 

I also don't have a use case for `hasFailure: false` right now so we could consider prohibiting that too.

Reply:
If the three cases above are mutually exclusive, would it make sense to use `oneOf` instead? I trust these rules are all tested with "invalid" tests so I won't dig too much into the JSON schema syntax here.

Since `data` is a required field and you mentioned `failure: { $$exists: true }` is equivalent with `hasFailure: true`, what does `hasFailure` give us apart from the ability to enforce use of `failureIsRedacted` in the schema?

Noting the following asks from https://github.com/mongodb/specifications/pull/1303/files#r981488414:

> 1. That an error code, error code name, and error labels are present/non-redacted.
> 2. That no other properties are present/unredacted.
> 3. That the attached server response only contains an error code, error code name, and error labels.

If there isn't a use case for `hasFailure: false` and `failureIsRedacted` is only allowed when `hasFailure: true`, I wonder why we can't stick to single `failureIsRedacted` boolean and make it a runtime error if the assertion field is used and there isn't a failure present (IIUC that would mean `data` omits a `failure` field).

Reply:
you're right that `oneOf` would make more sense here. 

> what does `hasFailure` give us apart from the ability to enforce use of `failureIsRedacted` in the schema?

nothing I can think of besides that. 

> I wonder why we can't stick to single `failureIsRedacted` boolean and make it a runtime error if the assertion field is used and there isn't a failure present (IIUC that would mean `data` omits a `failure` field).

this is close to what I initially had in the first place (the single boolean `hasFailure`), though better named. I would be fine with that approach, but switched to this since Jeff had suggested two booleans and I thought you were advocating for it as well in the thread below for schema enforcement purposes, hence the switch. that said, there was a lot of back and forth down there and I think the fact that `failure: { $$exists: true/false }` already covered failure presence/absence likely got lost in the mix.

wdyt @jyemin?

Reply:
ok, I've switched to @jmikola's suggestion of a single `failureIsRedacted` boolean plus having drivers check at runtime that it is only used in combination with `failure: { $$exists: true }`